ops/testing.py

```diff
@@ -1314,25 +1317,288 @@ def get_services(self, names: typing.List[str] = None) -> typing.List[pebble.Ser
+                # final directory.
+                # (At the time of writing, Pebble defaults to the specified permissions and
+                # root:root ownership, which is inconsistent with the push function's
+                # behavior for parent directories.)
+            else:
+                raise
+        if not isinstance(dir_, _Directory):
+            raise pebble.PathError(
+                'generic-file-error', 'parent is not a directory: {}'.format(str(dir_)))
+        return dir_.create_file(path_obj.name, data, encoding=encoding, **kwargs)
+
+    def list_dir(self, path) -> typing.List['_File']:
+        current_dir = self.root
+        tokens = pathlib.PurePosixPath(path).parts[1:]
+        for token in tokens:
+            try:
+                current_dir = current_dir[token]
+            except KeyError:
+                raise FileNotFoundError(str(current_dir.path / token))
+            if isinstance(current_dir, _File):
+                raise NotADirectoryError(str(current_dir.path))
+            if not isinstance(current_dir, _Directory):
+                # For now, ignoring other possible cases besides File and Directory (e.g. Symlink).
+                raise NotImplementedError()
+
+        return [child for child in current_dir]
+
+    def open(
+            self,
+            path: typing.Union[str, pathlib.PurePosixPath],
+            encoding: typing.Optional[str] = 'utf-8',
+    ) -> typing.Union[typing.BinaryIO, typing.TextIO]:
+        path = pathlib.PurePosixPath(path)
+        file = self[path]  # warning: no check re: directories
+        if isinstance(file, _Directory):
+            raise IsADirectoryError(str(file.path))
+        return file.open(encoding=encoding)
+
+    def __getitem__(self, path: typing.Union[str, pathlib.PurePosixPath]) \
+            -> typing.Union['_Directory', '_File']:
+        path = pathlib.PurePosixPath(path)
+        tokens = path.parts[1:]
+        current_object = self.root
+        for token in tokens:
+            # ASSUMPTION / TESTME: object might be file
+            if token in current_object:
+                current_object = current_object[token]
+            else:
+                raise FileNotFoundError(str(current_object.path / token))
+        return current_object
+
+    def __delitem__(self, path: typing.Union[str, pathlib.PurePosixPath]) -> None:
+        path = pathlib.PurePosixPath(path)
+        parent_dir = self[path.parent]
+        del parent_dir[path.name]
+
+
+class _Directory:
+    def __init__(self, path: pathlib.PurePosixPath, **kwargs):
+        self.path = path
+        self._children = {}
+        self.last_modified = datetime.datetime.now()
+        self.kwargs = kwargs
+
+    @property
+    def name(self) -> str:
+        # Need to handle special case for root.
+        # pathlib.PurePosixPath('/').name is '', but pebble returns '/'.
+        return self.path.name if self.path.name else '/'
+
+    def __contains__(self, child: str) -> bool:
```

Comment:
I'm on the fence about these -- but I think the "mapping magic" is reasonable here. A directory *is* more like a mapping (than a file system).

Reply:
At this point the _Directory class is the only bit that uses them now, and it is basically fronting a dictionary.  I'll leave it in here.